File: src/Commands.hs
module Commands
  ( Command(..)
  , readCommand
  ) where

data Command
  = TurnLeft
  | MoveForward
  | TurnRight
  deriving (Show, Eq)

readCommand :: Char -> Maybe Command
readCommand 'L' = Just TurnLeft
readCommand 'R' = Just TurnRight
readCommand 'F' = Just MoveForward
readCommand _   = Nothing

File: scala/src/main/scala/Commands.scala
package com.rumblesan.marsrover

object Commands {

  sealed trait Command
  case object TurnLeft extends Command
  case object MoveForward extends Command
  case object TurnRight extends Command

  def parse(c: Char): Option[Command] = {
    c match {
      case 'L' => Some(TurnLeft)
      case 'F' => Some(MoveForward)
      case 'R' => Some(TurnRight)
      case _ => None
    }
  }

}
